File: torch/precision.py
import os
import math
import torch
from export_vgtensor import export_vgtensor

DATA_DIR = "../vgrad/measurements/data"
os.makedirs(DATA_DIR, exist_ok=True)

torch.manual_seed(3)

N = 100
x = torch.randn(N, N)

export_vgtensor(x, os.path.join(DATA_DIR, "rand_matrix.vgtensor"))


for i in range(10):
    x = x @ x / math.sqrt(N)
    print(f"Variance: {x.var().item()}")

File: vgrad/measurements/precision.cpp
#include "vgrad.h"

using namespace vgrad;

int main() {
    using D = Dimension<100>;
    auto x = import_vgtensor<float, MakeShape<D, D>>("data/rand_matrix.vgtensor");

    for (int i = 0; i < 10; i++) {
        x = (matmul(x, x) / std::sqrt(D::value)).detach();
        auto m = mean(mean(x));
        auto var = mean(mean(pow(x - m, 2)));
        std::cout << "Variance: " << var.value() << std::endl;
    }
}